### PR TITLE
refactor(emergency_handler, mrm_handler) change MRM operators to classes inside Handler Nodes

### PR DESCRIPTION
**Pull request by veqcc**
## Description

Delete MRM operators and implemented those operators inside Handlers (Emergency Handler, MRM Handler) as classes based on https://github.com/autowarefoundation/autoware.universe/issues/6539 .

## Related links

https://github.com/autowarefoundation/autoware.universe/issues/6539

## Tests performed

Not yet.

## Notes for reviewers

Please review especially in terms of:
 - namespace structure
 - `update_rate` of handlers
   - Previously, it was 10 Hz in `emergency_handler`, while 33 Hz in `emergency_stop_operator`.  
 - how to pass parameters to each operator
 - When to publish emergency commands in emergency_stop operator class
   - only in `operate()` or both `operate()` and `onTimer()` ? 
 - Are commands `gear` and `hazard_light` used only when `emergency_stop` is enabled?
   - If so, should they included in `emergency_stop` class, not in handlers?

## Interface changes

We cannot access MRM operators directly.
All inputs/outputs of operators are included in the Handler's interfaces.

## Effects on system behavior

Handlers do not have to call ROS 2 services anymore to invoke mrm oprations.
It changes to function calls and thus mrm operations can work more quickly in terms of:
 - no need to wait next timer callback
 - no need to wait ROS 2 scheduler to schedule

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Review by isamu-takagi — COMMENTED** [review]
> namespace structure

Do you mean comfortable_stop_operator in emergency_handler? It looks fine.
FYI: It can be written as `namespace emergency_handler::comfortable_stop_operator`

> update_rate of handlers
> Previously, it was 10 Hz in emergency_handler, while 33 Hz in

Since it changes the publishing rate of control commands, please check whether it has any effect on vehicle control. Also, it is better to treat the topic rate and processing rate separately.

> emergency_stop_operator.
> how to pass parameters to each operator

Where is this about? I think it would be better to use `declare_parameter` in each operator class, as is the current implementation.

> When to publish emergency commands in emergency_stop operator class
> only in operate() or both operate() and onTimer() ?

I think it's better to always publish if there are no performance problems.

> Are commands gear and hazard_light used only when emergency_stop is enabled?
> If so, should they included in emergency_stop class, not in handlers?

Yes. At the meeting the other day, we decided to publish a command for each MRM, so please move it in the operator class. It should be needed for comfortable_stop as well.

File: system/emergency_handler/include/operators/comfortable_stop.hpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2024 TIER IV, Inc.
```
Comment (isamu-takagi) [inline system/emergency_handler/include/operators/comfortable_stop.hpp:1]: Place private headers that are not used by other packages under `src` directory.
https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/directory-structure/#include

File: system/emergency_handler/include/operators/emergency_stop.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
+// Copyright 2024 TIER IV, Inc.
```
Comment (isamu-takagi) [inline system/emergency_handler/include/operators/emergency_stop.hpp:1]: Place private headers that are not used by other packages under `src` directory.
https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/directory-structure/#include

File: system/emergency_handler/include/operators/emergency_stop.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  double target_jerk;
+};
+
+class EmergencyStopOperator
+{
+public:
+  explicit EmergencyStopOperator(rclcpp::Node * node);
+  bool operate();
+  bool cancel();
+  void onTimer();
```
Comment (isamu-takagi) [inline system/emergency_handler/include/operators/emergency_stop.hpp:42]: Timers should not be included in the interface, so each operator should create its own timer and change this function to private.

File: system/emergency_handler/include/operators/comfortable_stop.hpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  double min_jerk;
+};
+
+class ComfortableStopOperator
+{
+public:
+  explicit ComfortableStopOperator(rclcpp::Node * node);
+  bool operate();
+  bool cancel();
+  void onTimer();
```
Comment (isamu-takagi) [inline system/emergency_handler/include/operators/comfortable_stop.hpp:42]: Timers should not be included in the interface, so each operator should create its own timer and change this function to private.

**Comment by TetsuKawa** [comment]
As soon as this node is started up, the process dies.
I don't know where the bug is, but you should look into the code.

**Comment by veqcc** [comment]
@TetsuKawa 
Thank you for your comment.
Adding some parameters to top-level the configuration file works well.

**Comment by TetsuKawa** [comment]
@veqcc 
I'm sorry. You are right, the higher level parameter file was loaded.

**Comment by veqcc** [comment]
@isamu-takagi 
Thank you for your helpful comments!

> > namespace structure

> Do you mean comfortable_stop_operator in emergency_handler? It looks fine.
> FYI: It can be written as namespace emergency_handler::comfortable_stop_operator

Yes I meant it. I have fixed the namespace following it.

**Comment by veqcc** [comment]
> > emergency_stop_operator. how to pass parameters to each operator

> Where is this about? I think it would be better to use declare_parameter in each operator class, as is the current implementation.

Yes I meant the parameters in configuration files.
Thank you for your review.

**Comment by veqcc** [comment]
@isamu-takagi 

> > Are commands gear and hazard_light used only when emergency_stop is enabled? If so, should they included in emergency_stop class, not in handlers?

> Yes. At the meeting the other day, we decided to publish a command for each MRM, so please move it in the operator class. It should be needed for comfortable_stop as well.

I see. I will move these parameters and functions to operators from handlers.
One question here is, is it neccesary to modify `vehicle_cmd_gate` to enable these commands even when `comfortable_stop`?
It seems the `vehicle_cmd_gate` choosed emergency commands only when `emergency_stop` is being operated.
If so, I will make another Pull Request for it after this PR is merged.

File: system/emergency_handler/include/operators/emergency_stop.hpp (hunk 3 above)
Comment (veqcc) [reply system/emergency_handler/include/operators/emergency_stop.hpp]: :+1:

**Comment by veqcc** [comment]
> > update_rate of handlers
> > Previously, it was 10 Hz in emergency_handler, while 33 Hz in

> Since it changes the publishing rate of control commands, please check whether it has any effect on vehicle control. Also, it is better to treat the topic rate and processing rate separately.

I have changed following your advice. Now `emergency_handler` works in 10Hz, while `emergency_stop_operator` works in 30Hz as before.

**Comment by veqcc** [comment]
> > When to publish emergency commands in emergency_stop operator class. Only in operate() or both operate() and onTimer() ?

> I think it's better to always publish if there are no performance problems.

Both the current implementaion and my modification always publish them.
There are no performance problem in Psim on my computer, though I don't know in full Autoware on actual vehicles.
Until any performance problem happens, I will keep the current one.

**Comment by veqcc** [comment]
I willl close it.

**Pull request closed without merging**